eos/b-decays/b-to-3l-nu.cc

```diff
@@ -0,0 +1,187 @@
+/* vim: set sw=4 sts=4 et foldmethod=syntax : */
+
+/*
+ * Copyright (c) 2022 Danny van Dyk
+ * Copyright (c) 2022 Stephan Kürten
+ *
+ * This file is part of the EOS project. EOS is free software;
+ * you can redistribute it and/or modify it under the terms of the GNU General
+ * Public License version 2, as published by the Free Software Foundation.
+ *
+ * EOS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+ * Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <eos/b-decays/b-to-3l-nu.hh>
+#include <eos/form-factors/mesonic.hh>
+#include <eos/maths/complex.hh>
+#include <eos/maths/power-of.hh>
+#include <eos/models/model.hh>
+#include <eos/utils/destringify.hh>
+#include <eos/utils/options-impl.hh>
+#include <eos/utils/private_implementation_pattern-impl.hh>
+
+#include <string>
+#include <cmath>
+
+namespace eos
+{
+    using std::norm;
+    using std::abs;
+    using std::real;
+    using std::imag;
+    using std::sqrt;
+    using std::conj;
+    using std::log;
+
+    /*
+     * Decay: B_q^- -> lprime^+ lprime^- l^- nubar, cf. [KKvDZ:2022A]
+     */
+    template <>
+    struct Implementation<BToThreeLeptonsNeutrino>
+    {
+        std::shared_ptr<Model> model;
+
+        std::shared_ptr<FormFactors<PToGammaOffShell>> form_factors;
+
+        UsedParameter hbar;
+
+        UsedParameter g_fermi;
+
+        UsedParameter m_B;
+
+        UsedParameter f_B;
+
+        UsedParameter tau_B;
+
+        UsedParameter alpha_qed;
+
+        SwitchOption opt_lprime;
+
+        UsedParameter m_lprime;
+
+        SwitchOption opt_l;
+
+        UsedParameter m_l;
+
+        static const std::vector<OptionSpecification> options;
+
+        Implementation(const Parameters & p, const Options & o, ParameterUser & u) :
+            model(Model::make(o.get("model", "SM"), p, o)),
+            form_factors(FormFactorFactory<PToGammaOffShell>::create("B->gamma^*::" + o.get("form-factors", "KKvDZ2022"), p, o)),
+            hbar(p["QM::hbar"], u),
+            g_fermi(p["WET::G_Fermi"], u),
+            m_B(p["mass::B_u"], u),
+            f_B(p["decay-constant::B_u"], u),
+            tau_B(p["life_time::B_u"], u),
+            alpha_qed(p["QED::alpha_e(m_b)"],u),
+            opt_lprime(o, "lprime", {"e", "mu"}, "mu"),
+            m_lprime(p["mass::" + opt_lprime.value()], u),
+            opt_l(o, "l", {"e", "mu"}, "e"),
+            m_l(p["mass::" + opt_l.value()], u)
+        {
```

Comment:
Please add a check that ``l.value() != lprime.value()`` and throw an ``InvalidOptionValue`` for ``l`` otherwise.